### PR TITLE
sql: remove sql.defaults.optimizer_use_not_visible_indexes.enabled

### PR DESCRIPTION
**Pull request by michae2**
**sql: remove sql.defaults.optimizer_use_not_visible_indexes.enabled**

In 22.2 the use of cluster settings for session setting defaults is
deprecated, but I forgot this when reviewing #86033. Remove the
unreleased default cluster setting we added.

Release note: None

**sql: add more session variables to statement bundles**

Add some missing session variables to statement diagnostic bundles. This
isn't quite everything, but it's better than what we had before.

Release note: None

----

Release justification: low-risk bug fix.

### CODE REVIEW

**Review by mgartner — APPROVED** [review]
:lgtm: 

> Reviewed 6 of 6 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-:-NA9wIPI0zrqJ_BTDiCE:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2, @rafiss, @rytaft, and @wenyihu6)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-NA9sVMN8wwB-fU2hpFf):*
I don't think a release note is necessary because the setting is not included in any releases.
___
*[`pkg/sql/exec_util.go` line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-NA9siKM3tngi6EtT12Y:-NA9siKM3tngi6EtT12Z:b-p9dbnl) ([raw file](https://github.com/cockroachdb/cockroach/blob/869d44cb46ba1900202eb9e33a3353eecf06b269/pkg/sql/exec_util.go#L402)):*
> ```Go
> var optUseNotVisibleIndexesClusterMode = settings.RegisterBoolSetting(
> 	settings.TenantWritable,
> 	"sql.defaults.optimizer_use_not_visible_indexes.enabled",
> ```

Looks like the check added in #80575 isn't stern enough. Maybe we should add something like this to `pkg/settings/registry.go`:

```go
// WARNING: Nevertincrease this constant.
const maxSQLDefaults = 48;

func init() {
    if len(sqlDefaultSettings) > maxSQLDefaults {
        panic("new sql.defaults cluster settings are not needed now that `ALTER ROLE ... SET` syntax is supported")
    }
}
```

**Review by wenyihu6 — APPROVED** [review]
The first commit :lgtm: Thanks for fixing this!!

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-:-NA9yvuG8Z7-x7S3gYwj:bcaawlu)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @michae2, @rafiss, and @rytaft)

**Review by rafiss — CHANGES_REQUESTED** [review]
thanks for the change! let's not have a release note on this commit. a user reading the notes won't be able to make sense of it, since we never released a CRDB version that includes this cluster setting

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86634)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[`pkg/sql/exec_util.go` line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-NA9siKM3tngi6EtT12Y:-NAAEffV11N-tcEw9-Yd:b-wg44fp) ([raw file](https://github.com/cockroachdb/cockroach/blob/869d44cb46ba1900202eb9e33a3353eecf06b269/pkg/sql/exec_util.go#L402)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Looks like the check added in #80575 isn't stern enough. Maybe we should add something like this to `pkg/settings/registry.go`:

```go
// WARNING: Nevertincrease this constant.
const maxSQLDefaults = 48;

func init() {
    if len(sqlDefaultSettings) > maxSQLDefaults {
        panic("new sql.defaults cluster settings are not needed now that `ALTER ROLE ... SET` syntax is supported")
    }
}
```
</blockquote></details>

i don't quite understand why the existing check didn't work. do you know?

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86634)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[`pkg/sql/exec_util.go` line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-NA9siKM3tngi6EtT12Y:-NAAEji-8qQLkfLKRQmU:b-lle3by) ([raw file](https://github.com/cockroachdb/cockroach/blob/869d44cb46ba1900202eb9e33a3353eecf06b269/pkg/sql/exec_util.go#L402)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i don't quite understand why the existing check didn't work. do you know?
</blockquote></details>

using a constant for `maxSQLDefaults` could have an issue if a setting is retired and reduces the "expected" size of the allowed settings list

**Comment by wenyihu6** [comment]
*[`pkg/sql/exec_util.go` line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-NA9siKM3tngi6EtT12Y:-NAAH4dvDZkzrAtTjLAn:bx4cwqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/869d44cb46ba1900202eb9e33a3353eecf06b269/pkg/sql/exec_util.go#L402)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

using a constant for `maxSQLDefaults` could have an issue if a setting is retired and reduces the "expected" size of the allowed settings list
</blockquote></details>

That was my bad. I was following the existing pattern for other session variables and didn't realize that this has been deprecated. The commit got around the check by adding the session variable to this [map](https://github.com/cockroachdb/cockroach/blob/681f951100b1e5a1240c15a9841cd8f58b6b6c31/pkg/settings/registry.go#L180).   Any session variables added to this map [doesn't throw an error](https://github.com/cockroachdb/cockroach/blob/681f951100b1e5a1240c15a9841cd8f58b6b6c31/pkg/settings/registry.go#L212)

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86634)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner, @rytaft, and @wenyihu6)

___
*[`pkg/sql/exec_util.go` line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-NA9siKM3tngi6EtT12Y:-NAAL2xw9kPM1pjpH7lg:b-na5t37) ([raw file](https://github.com/cockroachdb/cockroach/blob/869d44cb46ba1900202eb9e33a3353eecf06b269/pkg/sql/exec_util.go#L402)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

That was my bad. I was following the existing pattern for other session variables and didn't realize that this has been deprecated. The commit got around the check by adding the session variable to this [map](https://github.com/cockroachdb/cockroach/blob/681f951100b1e5a1240c15a9841cd8f58b6b6c31/pkg/settings/registry.go#L180).   Any session variables added to this map [doesn't throw an error](https://github.com/cockroachdb/cockroach/blob/681f951100b1e5a1240c15a9841cd8f58b6b6c31/pkg/settings/registry.go#L212)
</blockquote></details>

no problem. then perhaps we can make the error message more clear so that someone else doesn't do the same thing.

**Comment by wenyihu6** [comment]
*[`pkg/sql/exec_util.go` line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-NA9siKM3tngi6EtT12Y:-NAAMM-z2yNwv6Kx9kcp:bhx1dh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/869d44cb46ba1900202eb9e33a3353eecf06b269/pkg/sql/exec_util.go#L402)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

no problem. then perhaps we can make the error message more clear so that someone else doesn't do the same thing.
</blockquote></details>

Adding comments on [top of this map](https://github.com/cockroachdb/cockroach/blob/681f951100b1e5a1240c15a9841cd8f58b6b6c31/pkg/settings/registry.go#L150) can also be useful.

**Review by michae2 — COMMENTED** [review]
TFTRs!

Removed the release note (instead I left a comment on https://github.com/cockroachdb/docs/issues/14819).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-:-NAAc3nSCcPabxIHW2tW:b-g30sqr)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @rafiss, @rytaft, and @wenyihu6)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-NA9sVMN8wwB-fU2hpFf):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't think a release note is necessary because the setting is not included in any releases.
</blockquote></details>

Done.
___
*[`pkg/sql/exec_util.go` line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86634#-NA9siKM3tngi6EtT12Y:-NAAcFBTBeuytWazpA0F:bwpam0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/869d44cb46ba1900202eb9e33a3353eecf06b269/pkg/sql/exec_util.go#L402)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Adding comments on [top of this map](https://github.com/cockroachdb/cockroach/blob/681f951100b1e5a1240c15a9841cd8f58b6b6c31/pkg/settings/registry.go#L150) can also be useful.
</blockquote></details>

I tried making the error message more clear, and added some comments.

**Review by rafiss — APPROVED** [review]
> Reviewed 3 of 6 files at r1, 2 of 5 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86634)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @rafiss, @rytaft, and @wenyihu6)

**Comment by michae2** [comment]
bors r=mgartner,wenyihu6,rafiss

**Comment by michae2** [comment]
Trying again.

bors r=mgartner,wenyihu6,rafiss

File: pkg/settings/registry.go [hunk 1]
```diff
@@ -147,7 +147,12 @@ var retiredSettings = map[string]struct{}{
 	"sql.defaults.intervalstyle.enabled":                        {},
 }
 
+// sqlDefaultSettings is the list of "grandfathered" existing sql.defaults
+// cluster settings. In 22.2 and later, new session settings do not need an
+// associated sql.defaults cluster setting. Instead they can have their default
+// changed with ALTER ROLE ... SET.
 var sqlDefaultSettings = map[string]struct{}{
+	// PLEASE DO NOT ADD NEW SETTINGS TO THIS MAP. THANK YOU.
```
Comment (mgartner) [inline pkg/settings/registry.go:155]: 🖖 

Thanks for adding these warnings!

**Comment by michae2** [comment]
Third time's the charm.

bors r=mgartner,wenyihu6,rafiss

**Comment by michae2** [comment]
"It's definitely going to work this time."

bors r=mgartner,wenyihu6,rafiss

**Comment by mgartner** [comment]
The second commit makes progress towards #72434. @michae2 do you think there are other session settings that we should add to the statement bundle, or should we just close #72434?

**Comment by michae2** [comment]
> The second commit makes progress towards #72434. @michae2 do you think there are other session settings that we should add to the statement bundle, or should we just close #72434?

This was just a quick-and-dirty attempt to add more while I was in the area. I think I probably missed a few, and something more systematic (e.g. automatically adding all session settings) would be better. So I left the TODO in the code, and I think we should leave #72434 open.

**Pull request merged**